### ISSUE TITLE
Field company_id not required in res.currency, but should be in multicompany

### ISSUE THREAD

**Issue by bealdav** [odoo/odoo#641]
Currencies can be implicitly shared between companies but it produces errors on currency rates

**_Impacted versions:**_
Odoo V7
Odoo V8

**_Steps to reproduce:**_

Field company_id in 'res.currency' is not mandatory (menu accounting / config / miscellanous / currencies). 

**_Current behavior:**_
You can share currencies between companies. In this case 'res.currency.rate' are completely false in these cases.
- Your companies have the same default currency but they don't want share the currency rates
- Your companies have a different default currency and then currency rates are incompatible

**_Expected behavior:**_
- Either company_id field in 'res.currency' is set to required and defined to the first company. Currencies are explicitly not shared
- Either company_id field is moved from 'res.currency' to 'res.currency.rate' : 
  In this case currencies are shared. It makes sense because currencies are defined for the whole world. Default currency can be different by company. Rate change (and update frequencies) will be modified over the time with a full reliability of the datas. (This is my prefered scenario but with more modification on the data model)

**Comment by pedrobaeza** [comment]
I think the most useful approach is to move company_id field to res.currency.rate, so that each company can have different rates (or not). Unshare currencies then doesn't matter, because you want to unshare it due to the lack of the company field in the other table.

**Comment by sebastienbeau** [comment]
I agree Pedro having share currency rate will be a better solution (currency are internationnal and it make sense to share the currency). Moreover it's seem simple to implement and I pretty that most of module will be compatible without any change (the rate will be filtred by the acces rule)

**Comment by rvalyi** [comment]
agree too: just having a different rate per company will do the job better.

**Comment by odony** [comment]
Could you please explain the use case that you think is not possible with the current model? What do you mean with "produces errors on currency rates"? What prevents correcting those errors with a proper configuration of the multi-company currencies (e.g. different currencies for each company, depending on your requirements)

When multicompany currencies were designed there were 2 options: putting the `company_id` field on `res.currency`, or putting it on `res.currency.rate`. Both solutions are functionally equivalent in the sense that they allow the same flexibility. It was decided to put it on the currency because that gives a more consistent "company context" around currencies. There might be cases where extra company-specific data needs to be added to currencies, so having the separation at the currency record is cleaner. In addition, when you're working as the admin (UID 1) you aren't affected by "record rules", so managing a random mix of rates on the currency form would be more error-prone, and the value for "current rate" would be semi-random.
When currencies are different for each company the separation is cleaner, and the risk of confusion is reduced for the admin (the admin account is typically not used for regular transactions where selecting a currency is necessary - as a best practice it is used mostly for configuration).

Regarding the "sharing of data": allowing full sharing is as important as allowing no sharing at all, because multi-company requirements are very different, each deployment is quite unique for this aspect. And this is true regardless of the model on with the `company_id` field is located.

To sum it up:
 - we explicitly **want** the **option** to share data between companies (even the rates), so the first item of your "expected behavior" is not valid
 - having the `company_id` field on the rate model is functionally equivalent and has no clear advantage over the current model (it's a trade-off of pros and cons), so I see no compelling reason to change it
 - yes "currencies are defined for the whole world", but their usage is different for each business, and defining them for each company makes sense too (the default list could even be reduced to only the currencies that each company cares about)

**Comment by sebastienbeau** [comment]
Hi Olivier
I can give you an example. Akretion use Euro and Real for his currency.

If we want to have it working with the actual data model, we need to create 2 euro currency and 2 real currency (we can not share it). Indeed if we share it the rate are share too, and as the currency of each company are not the same then the rate must be different (for akretion brasil the rate 0,32 (real=>euro) and for akretion france the rate is 3.06 (euro=>real).
This work but I think it's a pitty to not have the posibility to share the currency (my point of view). It's make harder the automatisation of invoicing between company, the possibility to share a pricelist, and also the set up of a new company (we need to recreate all of the currency needed).
And also as the company_id is not mandatory, user can make the error to have a sharred currency and accounting error when there have 2 company that have a different currency.

I would prefer a solution where the company will be on the currency and on the currency.rate.
With that case I can share the currency euro and real, but not share for exemple not share the Peso Argentino as we do not use it for the French Part. 
An alternative way will to have a company_ids on the currency (each company can choose the currency needed) and have a company rate on the res.currency.rate

What do you think?

**Comment by pedrobaeza** [comment]
Another use case in favor of having company_id at currency rate level: two companies have the same base currency, and want to share it, but they make currency rate freezes (one uses a fixed rate per month, and the other uses daily rate). This can only be achieved adding this field.

**Comment by gurneyalex** [comment]
On 15/01/2015 12:55, Olivier Dony wrote:
>
> Could you please explain the use case that you think is not possible
> with the current model? What do you mean with "produces errors on
> currency rates"? What prevents correcting those errors with a proper
> configuration of the multi-company currencies (e.g. different
> currencies for each company, depending on your requirements)
>
> When multicompany currencies were designed there were 2 options:
> putting the |company_id| field on |res.currency|, or putting it on
> |res.currency.rate|. Both solutions are functionally equivalent in the
> sense that they allow the same flexibility. It was decided to put it
> on the currency because that gives a more consistent "company context"
> around currencies. There might be cases where extra company-specific
> data needs to be added to currencies, so having the separation at the
> currency record is cleaner. In addition, when you're working as the
> admin (UID 1) you aren't affected by "record rules", so managing a
> random mix of rates on the currency form would be more error-prone,
> and the value for "current rate" would be semi-random.
> When currencies are different for each company the separation is
> cleaner, and the risk of confusion is reduced for the admin (the admin
> account is typically not used for regular transactions where selecting
> a currency is necessary - as a best practice it is used mostly for
> configuration).
>
> Regarding the "sharing of data": allowing full sharing is as important
> as allowing no sharing at all, because multi-company requirements are
> very different, each deployment is quite unique for this aspect. And
> this is true regardless of the model on with the |company_id| field is
> located.
>
> To sum it up:
>
>   * we explicitly *want* the *option* to share data between companies
>     (even the rates), so the first item of your "expected behavior" is
>     not valid
>   * having the |company_id| field on the rate model is functionally
>     equivalent and has no clear advantage over the current model (it's
>     a trade-off of pros and cons), so I see no compelling reason to
>     change it
>   * yes "currencies are defined for the whole world", but their usage
>     is different for each business, and defining them for each company
>     makes sense too (the default list could even be reduced to only
>     the currencies that each company cares about)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/odoo/odoo/issues/641#issuecomment-70075496>.
>
Hello Olivier,

What we don't want is
https://bugs.launchpad.net/openobject-addons/+bug/1111298 comming back
to haunt us.

ALexandre

-- 
Alexandre Fayolle
Chef de Projet
Tel : + 33 (0)4 79 26 57 94

Camptocamp France SAS
Savoie Technolac, BP 352
73377 Le Bourget du Lac Cedex
http://www.camptocamp.com

**Comment by odony** [comment]
@sebastienbeau: Thanks for your answer, it confirms that everything is possible as long as you configure it properly. As for what I think of your proposed solution, I've explained in my previous comment why it's not actually better than the current situation, and I still see no compelling reason to change.
Let me stress once more that multi-company setup is always more complex than it seems, every case has very different requirements, and for every example in favor of one model, someone else can come up with an opposite example in favor of the other model.

@gurneyalex that's one of the reasons why changing the model without a really good reason is not a good idea

@pedrobaeza in your example "have the same currency and want to share it" simply makes no functional or business sense. The business requirement is either that they want to keep the currencies and rates in sync or they don't, period. The rest is just a technical configuration detail, as users in each company will only ever see a single instance of each currency in either case.

**Comment by pedrobaeza** [comment]
@odony, we are talking here about comfort. Of course you can configure a currency for each company, and the rest of the propagated data that links to a currency, but you have two problems:
* This means a lot of work. As @sebastienbeau said, it means to duplicate a lot of records. My example goes in the same direction. Why duplicate a lot of things, when you can only duplicate what changes (rates)?
* A potential problem that you define a model with a currency link that it's not multi-company aware. In this case, you are forced to duplicate each record with one currency, which are seen for all companies.

**Comment by sebastienbeau** [comment]
After thinking twice, I think I have one case that can not work with the current data model.
Indeed if you are a group of company and you have a contract with a customer so the pricing is the same in all of the company then you may want to share the pricelist (instead of duplicating it). But sharing a pricelist is only posible if you can share the currency or in case of multi-company which didn't have the same main currency you can not share a currency.
For sure we can duplicate the pricelist in each company, so we can always found a work around. And no need to change the data model in the core

I think we can try to manage this in a community module

@pedrobaeza what do you think about doing a community module that add the company on res.currency.rate ? We can do it.

**Comment by pedrobaeza** [comment]
@sebastienbeau Yes, why not? But it's a pity that we can't get this into core.

**Comment by sebastienbeau** [comment]
yes, it's a pity but I think we can do it without compatibilty trouble so let's do it ;)

**Comment by pedrobaeza** [comment]
This is the oldest opened issue right now, but... good news: this was finally implemented in the v9 accounting change following my suggestion! (see https://github.com/odoo/odoo/commit/c04065abd8f62c9a211c8fa824f5eecf68e61b73#diff-eedcb99c2a3b66f9db796c8b09a30bcaR280), so this can be closed.
